File: mindspore/ccsrc/runtime/device/tensor_array.cc
/**
 * Copyright 2021-2022 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "runtime/device/tensor_array.h"

namespace mindspore {
namespace device {
bool TensorArray::CheckValue(const TypeId &dtype, const ShapeVector &shape) {
  MS_LOG(DEBUG) << "Check the data shape and type for " << name_;
  MS_EXCEPTION_IF_NULL(dtype_);
  if (dtype != dtype_->type_id()) {
    MS_LOG(ERROR) << "Invalid data type " << TypeIdLabel(dtype) << " for " << name_ << ", the origin type is "
                  << TypeIdLabel(dtype_->type_id());
    return false;
  }
  if (shape != shapes_) {
    MS_LOG(ERROR) << "Invalid data shape " << shape << " for " << name_ << ", the origin shape is " << shapes_;
    return false;
  }
  return true;
}

bool TensorArray::CheckReadIndexLogical(const int64_t index) {
  if (LongToSize(index) >= valid_size_) {
    MS_LOG(ERROR) << "Index " << index << " out of range " << valid_size_ << ", " << name_;
    return false;
  }
  return true;
}

// Function Read() can get the tensors in the scope of tensors_.
mindspore::kernel::AddressPtr TensorArray::Read(const int64_t index) {
  if (LongToSize(index) >= tensors_.size()) {
    MS_LOG(EXCEPTION) << "Index " << index << " out of range " << tensors_.size() << ", " << name_;
  }
  MS_LOG(DEBUG) << "Read tensor index = " << index << ", addr = " << tensors_[LongToSize(index)]->addr;
  return tensors_[LongToSize(index)];
}

// Add tensor to the TensorArray and increase the size.
// Cast 1: is_dynamic = False and index > max_size_, error.
// Case 2: index > valid_size, fill the rest dev_value with zeros, and set valid_size to index + 1.
// Case 3: index == tensors_.size(), we need to increase both real tensors_ size and valid size, and add
// the new dev_value to tensors_.
// Case 4: tensors_size() > index > valid_size, we can reuse the memory in tensors_[index], so
// only increase the valid_size.
bool TensorArray::Write(const int64_t index, const mindspore::kernel::AddressPtr &dev_value) {
  MS_LOG(DEBUG) << "Write dev_value to " << name_;
  if (!is_dynamic_ && (index >= max_size_)) {
    MS_LOG(ERROR) << name_ << " is not in dynamic size, the max_size is " << max_size_ << ", but get index " << index;
    return false;
  }
  if (LongToSize(index) > valid_size_) {
    // Create/reuse (index - valid_size) size dev_value with zeros.
    // 1 create new mem : index > real_size ? index - real_size : 0
    // 2 reuse old mem : index > real_size ? real_size - valid_size : index - valid_size
    // 3 fill zeros : index - valid_size
    MS_EXCEPTION_IF_NULL(dev_value);
    size_t create_size = (LongToSize(index) > tensors_.size()) ? (LongToSize(index) - tensors_.size()) : 0;
    for (size_t i = 0; i < create_size; i++) {
      kernel::AddressPtr create_dev = std::make_shared<kernel::Address>();
      create_dev->addr = AllocateMemory(dev_value->size);
      create_dev->size = dev_value->size;
      tensors_.push_back(create_dev);
    }
    tensors_.push_back(dev_value);
    for (size_t i = valid_size_; i < LongToSize(index); i++) {
      MS_EXCEPTION_IF_CHECK_FAIL((tensors_.size() > i), "The index is out of range.");
      MS_EXCEPTION_IF_NULL(tensors_[i]);
      auto tensor_size = tensors_[i]->size;
      ClearMemory(tensors_[i]->addr, tensor_size);
    }
    valid_size_ = LongToSize(index) + 1;
  } else if (LongToSize(index) == tensors_.size()) {
    MS_LOG(DEBUG) << "Write to index " << index << ", increase tensors' size to " << (tensors_.size() + 1);
    tensors_.push_back(dev_value);
    valid_size_++;
  } else {
    MS_LOG(DEBUG) << "Reuse tensors in position " << index << ", tensors size is " << tensors_.size();
    if (LongToSize(index) == valid_size_) {
      valid_size_++;
    }
  }
  return true;
}

void TensorArray::Clear() {
  valid_size_ = 0;
  return;
}

void TensorArray::Free() {
  MS_LOG(DEBUG) << "Free device memory for " << name_;
  for (const auto &addr : tensors_) {
    if (addr != nullptr) {
      FreeMemory(static_cast<DeviceMemPtr>(addr->addr));
    }
  }
}

size_t TensorArray::GetValidSize() const { return valid_size_; }
size_t TensorArray::GetRealSize() const { return tensors_.size(); }

const void *TensorArray::GetTensorAddr(const size_t &index) const {
  MS_EXCEPTION_IF_CHECK_FAIL((tensors_.size() > index), "The index is out of range.");
  MS_EXCEPTION_IF_NULL(tensors_[index]);
  return tensors_[index]->addr;
}

void TensorArray::SetMaxSize(const int64_t size, const bool is_dynamic) {
  is_dynamic_ = is_dynamic;
  if (!is_dynamic_) {
    max_size_ = size;
    MS_LOG(DEBUG) << name_ << " use fixed size " << max_size_;
  }
  return;
}
}  // namespace device
}  // namespace mindspore

File: mindspore/python/mindspore/ops/operations/_tensor_array.py
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

"""Operators for TensorArray."""

import mindspore as ms
from mindspore import _checkparam as validator
from ...common import dtype as mstype
from ..primitive import prim_attr_register, PrimitiveWithInfer, Primitive


class TensorArray(PrimitiveWithInfer):
    r"""
    TensorArrayCreate used to create a TensorArray and return an unique handle.

    .. warning::
        This is an experimental API that is subject to change or deletion.

    Args:
        dtype (mindspore.dtype): the data type in the TensorArray.
        element_shape (tuple[int]): the shape of each tensor in a TensorArray.
        dynamic_size (bool): If true the TensorArray can increase the size. Default: ``True``.
        size (int): The size of the TensorArray if dynamic_size = False.
        name (string): the name of this TensorArray. Default: "TA".

    Inputs:
        None.

    Outputs:
        - **output** (Tensor[mindspore.int64]) - an unique handle binded to the TensorArray.

    Supported Platforms:
        ``GPU`` ``CPU``

    Examples:
        >>> import mindspore
        >>> import mindspore.ops as ops
        >>> create_op = ops.TensorArray(mindspore.int32, ())
        >>> handle = create_op()
        >>> print(handle)
        0
    """
    @prim_attr_register
    def __init__(self, dtype, element_shape, dynamic_size=True, size=0, name="TA"):
        validator.check_type_name("dtype", dtype, mstype.number_type + (mstype.bool_,), self.name)
        validator.check_int(size, 0, validator.GE, "size", self.name)
        self.add_prim_attr('dtype', dtype)
        self.add_prim_attr('element_shape', element_shape)
        self.add_prim_attr('dynamic_size', dynamic_size)
        self.add_prim_attr('size', size)
        self.add_prim_attr('side_effect_mem', True)
        self.add_prim_attr('name', name)

    def infer_shape(self):
        return ()

    def infer_dtype(self):
        return mstype.int64


class TensorArrayWrite(PrimitiveWithInfer):
    r"""
    TensorArrayWrite used to write tensor into a created TensorArray.

    .. warning::
        This is an experimental API that is subject to change or deletion.

    Inputs:
        - **index** (Tensor[int64]) - The position to write.
        - **value** (Tensor) - The value to add into the TensorArray.
        - **handle** (Tensor[int64]) - The handle pointed to the TensorArray.

    Outputs:
        None.

    Supported Platforms:
        ``GPU`` ``CPU``

    Examples:
        >>> import mindspore
        >>> import mindspore.ops as ops
        >>> create_op = ops.TensorArray(mindspore.int32, ())
        >>> handle = create_op()
        >>> write_op = ops.TensorArrayWrite()
        >>> write_op.write(handle, 0, 1)
    """
    @prim_attr_register
    def __init__(self):
        self.add_prim_attr('side_effect_mem', True)

    def infer_shape(self, handle_shape, index_shape, value_shape):
        return ()

    def infer_dtype(self, handle_type, index_type, value_type):
        validator.check_type_name("handle", handle_type, (ms.int64), self.name)
        validator.check_type_name("index", index_type, (int, ms.int64), self.name)
        validator.check_type_name("value", value_type, mstype.number_type + (mstype.bool_,), self.name)
        return mstype.int64


class TensorArrayRead(PrimitiveWithInfer):
    r"""
    TensorArrayRead used to read tensor from a created TensorArray by the given index.

    .. warning::
        This is an experimental API that is subject to change or deletion.

    Args:
        dtype (mindspore.dtype): the data type in the TensorArray.
        element_shape (tuple[int]): the shape of each tensor in a TensorArray.

    Inputs:
        - **index** (Tensor[int64]) - The position to read.
        - **handle** (mindspore.int64) - The handle pointed to the TensorArray.

    Outputs:
        - **output** (Tensor) - the value in position index.

    Supported Platforms:
        ``GPU`` ``CPU``

    Examples:
        >>> import mindspore
        >>> import mindspore.ops as ops
        >>> create_op = ops.TensorArray(mindspore.int32, ())
        >>> handle = create_op()
        >>> write_op = ops.TensorArrayWrite()
        >>> write_op.write(handle, 0, 1)
        >>> read_op = ops.TensorArrayRead(mindspore.int32, ())
        >>> ans = read_op(handle, 0)
        >>> print(ans)
        1
    """
    @prim_attr_register
    def __init__(self, dtype, element_shape):
        validator.check_type_name("dtype", dtype, mstype.number_type + (mstype.bool_,), self.name)
        self.add_prim_attr('dtype', dtype)
        self.add_prim_attr('element_shape', element_shape)
        self.add_prim_attr('side_effect_mem', True)
        self.dtype = dtype
        self.shape = element_shape

    def infer_shape(self, handle_shape, index_shape):
        return self.shape

    def infer_dtype(self, handle_type, index_type):
        validator.check_type_name("handle", handle_type, (ms.int64), self.name)
        validator.check_type_name("index", index_type, (int, ms.int64), self.name)
        return self.dtype


class TensorArrayClose(PrimitiveWithInfer):
    r"""
    TensorArrayClose used to close the created TensorArray. The resources in TensorArray will be deleted.

    .. warning::
        This is an experimental API that is subject to change or deletion.

    Inputs:
        - **handle** (mindspore.int64) - The handle pointed to the TensorArray.

    Outputs:
        None.

    Supported Platforms:
        ``GPU`` ``CPU``

    Examples:
        >>> import mindspore
        >>> import mindspore.ops as ops
        >>> create_op = ops.TensorArray(mindspore.int32, ())
        >>> handle = create_op()
        >>> close_op = ops.TensorArrayClose()
        >>> close_op(handle)
    """
    @prim_attr_register
    def __init__(self):
        self.add_prim_attr('side_effect_mem', True)

    def infer_shape(self, handle_shape):
        return ()

    def infer_dtype(self, handle_type):
        validator.check_type_name("handle", handle_type, (ms.int64), self.name)
        return mstype.int64


class TensorArrayClear(PrimitiveWithInfer):
    r"""
    TensorArrayClear used to reset the created TensorArray. The instance of TensorArray is still aviliable.

    .. warning::
        This is an experimental API that is subject to change or deletion.

    Inputs:
        - **handle** (mindspore.int64) - The handle pointed to the TensorArray.

    Outputs:
        None.

    Supported Platforms:
        ``GPU`` ``CPU``

    Examples:
        >>> import mindspore
        >>> import mindspore.ops as ops
        >>> create_op = ops.TensorArray(mindspore.int32, ())
        >>> handle = create_op()
        >>> clear_op = ops.TensorArrayClear()
        >>> clear_op(handle)
    """
    @prim_attr_register
    def __init__(self):
        self.add_prim_attr('side_effect_mem', True)

    def infer_shape(self, handle_shape):
        return ()

    def infer_dtype(self, handle_type):
        validator.check_type_name("handle", handle_type, (ms.int64), self.name)
        return mstype.int64


class TensorArrayStack(Primitive):
    r"""
    TensorArrayStack used to stack the tensors in a created TensorArray into one tensor.

    .. warning::
        This is an experimental API that is subject to change or deletion.

    Args:
        dtype (mindspore.dtype): the data type in the TensorArray.
        element_shape (tuple[int]): the shape of each tensor in a TensorArray.

    Inputs:
        - **handle** (mindspore.int64) - The handle pointed to the TensorArray.

    Outputs:
        - **output** (Tensor) - the stacked value from the TensorArray.

    Supported Platforms:
        ``GPU`` ``CPU``

    Examples:
        >>> import mindspore
        >>> import mindspore.ops as ops
        >>> create_op = ops.TensorArray(mindspore.int32, ())
        >>> handle = create_op()
        >>> write_op = ops.TensorArrayWrite()
        >>> write_op.write(handle, 0, 1)
        >>> write_op.write(handle, 1, 2)
        >>> stack_op = ops.TensorArrayStack(mindspore.int32, ())
        >>> ans = stack_op(handle)
        >>> print(ans)
        [1 2]
    """
    @prim_attr_register
    def __init__(self, dtype, element_shape, dynamic_size, size):
        """Initialize TensorArrayStack"""
        self.init_prim_io_names(inputs=[''], outputs=['output'])
        self.add_prim_attr('dtype', dtype)
        self.add_prim_attr('element_shape', element_shape)
        self.add_prim_attr('is_dynamic_shape', dynamic_size)
        self.add_prim_attr('size', size)
        self.add_prim_attr('side_effect_mem', True)


class TensorArraySize(PrimitiveWithInfer):
    r"""
    TensorArraySize used to get the logical size of the created TensorArray.

    .. warning::
        This is an experimental API that is subject to change or deletion.

    Inputs:
        - **handle** (mindspore.int64) - The handle pointed to the TensorArray.

    Outputs:
        - **output** (Tensor[mindspore.int64]) - the logical size of the TensorArray.

    Supported Platforms:
        ``GPU`` ``CPU``

    Examples:
        >>> import mindspore
        >>> import mindspore.ops as ops
        >>> create_op = ops.TensorArray(mindspore.int32, ())
        >>> handle = create_op()
        >>> size_op = ops.TensorArraySize()
        >>> size = size_op(handle)
    """
    @prim_attr_register
    def __init__(self):
        self.add_prim_attr('side_effect_mem', True)

    def infer_shape(self, handle_shape):
        return ()

    def infer_dtype(self, handle_type):
        validator.check_type_name("handle", handle_type, (ms.int64), self.name)
        return mstype.int64


class TensorArrayGather(PrimitiveWithInfer):
    r"""
    TensorArrayGather used to gather specified elements from the created TensorArray.

    .. warning::
        This is an experimental API that is subject to change or deletion.

    Args:
        dtype (mindspore.dtype): the data type in the TensorArray.
        element_shape (tuple[int]): the shape of each tensor in a TensorArray.

    Inputs:
        - **handle** (mindspore.int64) - The handle pointed to the TensorArray.
        - **indices** (mindspore.int32) - The locations of the gathered elements.

    Outputs:
        - **output** (Tensor) - The gathered value from the TensorArray.

    Examples:
        >>> import mindspore
        >>> import mindspore.ops as ops
        >>> from mindspore import numpy as mnp
        >>> create_op = ops.TensorArray(mindspore.float32, dynamic_size=False, element_shape=(8,))
        >>> handle = create_op()
        >>> indices = mnp.range(0, 25, 1, mindspore.int32)
        >>> gather_op = ops.TensorArrayGather(dtype=mindspore.float32, element_shape=(8,))
        >>> gather_result = gather_op(handle, indices)
    """
    @prim_attr_register
    def __init__(self, dtype, element_shape):
        self.init_prim_io_names(inputs=['handle', 'indices'], outputs=['value'])
        self.add_prim_attr("side_effect_mem", True)
        self.dtype = dtype
        self.element_shape = element_shape

    def infer_shape(self, handle, indices):
        if len(indices) != 1:
            return ValueError("indices dimension should be equal to 1")
        return [indices[0]] + list(self.element_shape)

    def infer_dtype(self, handle, indices):
        validator.check_type_name("handle", handle, (ms.int64), self.name)
        validator.check_type_name("indices", indices, (ms.int32), self.name)
        return self.dtype
